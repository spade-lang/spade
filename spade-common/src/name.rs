use serde::{Deserialize, Serialize};

use crate::location_info::{Loc, WithLocation};

#[derive(PartialEq, Debug, Clone, Eq, Hash, Serialize, Deserialize)]
pub struct Identifier(pub String);

impl std::fmt::Display for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(PartialEq, Debug, Clone, Eq, Hash, Serialize, Deserialize)]
pub struct Path(pub Vec<Loc<Identifier>>);

impl Path {
    pub fn as_strs(&self) -> Vec<&str> {
        self.0.iter().map(|id| id.inner.0.as_ref()).collect()
    }
    pub fn as_strings(&self) -> Vec<String> {
        self.0.iter().map(|id| id.inner.0.clone()).collect()
    }
    /// Generate a path from a list of strings
    pub fn from_strs(elems: &[&str]) -> Self {
        Path(
            elems
                .iter()
                .map(|x| Identifier(x.to_string()).nowhere())
                .collect(),
        )
    }

    pub fn ident(ident: Loc<Identifier>) -> Self {
        Self(vec![ident])
    }

    pub fn push_ident(&self, ident: Loc<Identifier>) -> Path {
        let mut result = self.clone();
        result.0.push(ident);
        result
    }

    pub fn pop(&self) -> Self {
        let mut result = self.clone();
        result.0.pop().expect("Failed to pop identifier from path");
        result
    }

    pub fn join(&self, other: Path) -> Path {
        let mut result = self.clone();
        for ident in other.0 {
            result = result.push_ident(ident);
        }
        result
    }

    /// If the path is lib::<rest> return Some(<rest>), else None
    pub fn lib_relative(&self) -> Option<Path> {
        if self.0.first() == Some(&Identifier("lib".to_string()).nowhere()) {
            Some(Path(Vec::from(&self.0[1..])))
        } else {
            None
        }
    }

    /// The last element of the path. Panics if the path is empty
    pub fn tail(&self) -> Identifier {
        self.0
            .last()
            .expect("Tried getting tail of empty path")
            .inner
            .clone()
    }

    /// Returns the whole path apart from the tail. Panics if the path is empty
    pub fn prelude(&self) -> Path {
        Self(self.0[0..self.0.len() - 1].to_owned())
    }
}

impl std::fmt::Display for Path {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_strs().join("::"))
    }
}

/// Anything named will get assigned a unique name ID during AST lowering in order to avoid caring
/// about scopes once HIR has been generated. This is the type of those IDs
///
/// The associated string is only used for formatting when printing. The hash and eq methods do not
/// use it
#[derive(Clone, Serialize, Deserialize)]
pub struct NameID(pub u64, pub Path);

impl std::cmp::PartialEq for NameID {
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}

impl std::cmp::Eq for NameID {}

impl std::cmp::PartialOrd for NameID {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.0.cmp(&other.0))
    }
}

impl std::cmp::Ord for NameID {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.0.cmp(&other.0)
    }
}

impl std::hash::Hash for NameID {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

impl std::fmt::Debug for NameID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}#{}", self.1, self.0)
    }
}
impl std::fmt::Display for NameID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.1)
    }
}

pub mod testutil {
    use super::*;
    pub fn name_id(id: u64, name: &str) -> Loc<NameID> {
        NameID(id, Path::from_strs(&[name])).nowhere()
    }

    /// Shorthand for creating a name_id with static strs as name
    pub fn name_id_p(id: u64, name: &[&str]) -> Loc<NameID> {
        NameID(id, Path::from_strs(name)).nowhere()
    }
}
