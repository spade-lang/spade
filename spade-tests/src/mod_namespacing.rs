use crate::snapshot_error;
use spade_common::name::Path;

snapshot_error! {
    missing_mod_for_file,
    {
        {
            Path::from_strs(&["proj"]),
            Path::from_strs(&["proj"]),
            "src/main.spade",
            "fn just_some_stuff() {}"
        },
        {
            Path::from_strs(&["proj", "subfile"]),
            Path::from_strs(&["proj"]),
            "src/subfile.spade",
            ""
        }
    },
    false
}

snapshot_error! {
    missing_mod_for_directory,
    {
        {
            Path::from_strs(&["proj"]),
            Path::from_strs(&["proj"]),
            "src/main.spade",
            "fn just_some_stuff() {}"
        },
        {
            Path::from_strs(&["proj", "submod"]),
            Path::from_strs(&["proj"]),
            "src/submod/mod.spade",
            ""
        }
    },
    false
}

snapshot_error! {
    missing_mod_for_subfile_in_subdir,
    {
        {
            Path::from_strs(&["proj"]),
            Path::from_strs(&["proj"]),
            "src/main.spade",
            "mod submod;"
        },
        {
            Path::from_strs(&["proj", "submod"]),
            Path::from_strs(&["proj"]),
            "src/submod/main.spade",
            ""
        },
        {
            Path::from_strs(&["proj", "submod", "subfile"]),
            Path::from_strs(&["proj", "submod"]),
            "src/submod/subfile.spade",
            ""
        }
    },
    false
}

snapshot_error! {
    mod_but_no_file,
    {
        {
            Path::from_strs(&["proj"]),
            Path::from_strs(&["proj"]),
            "src/main.spade",
            "mod submod;
            "
        },
        {
            Path::from_strs(&["proj", "submod"]),
            Path::from_strs(&["proj"]),
            "src/submod/main.spade",
            "mod subsubmod;
            "
        }
    },
    false
}

snapshot_error! {
    missing_mod_in_external_lib,
    {
        {
            Path::from_strs(&["proj"]),
            Path::from_strs(&["proj"]),
            "src/main.spade",
            "mod submod;
            "
        },
        {
            Path::from_strs(&["dep"]),
            Path::from_strs(&["dep"]),
            "build/test/src/main.spade",
            "mod submod;
            "
        },
    },
    false
}
