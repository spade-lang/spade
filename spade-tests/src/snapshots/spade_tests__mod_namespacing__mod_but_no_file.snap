---
source: spade-tests/src/mod_namespacing.rs
---
mod submod;
            
mod subsubmod;
            

error: Did not find src/submod/subsubmod.spade or src/submod/subsubmod/main.spade
  ┌─ src/submod/main.spade:1:5
  │
1 │ mod subsubmod;
  │     ^^^^^^^^^ Did not find file for subsubmod
